horovod/common/ops/mpi_operations.h

```diff
@@ -0,0 +1,193 @@
+// Copyright 2016 The TensorFlow Authors. All Rights Reserved.
+// Modifications copyright (C) 2019 Uber Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// =============================================================================
+
+#ifndef HOROVOD_MPI_OPERATIONS_H
+#define HOROVOD_MPI_OPERATIONS_H
+
+#include <iostream>
+#include <queue>
+
+#include "mpi.h"
+
+#include "../common.h"
+#include "../communication_channel.h"
+#include "../global_state.h"
+#include "collective_operations.h"
+
+#if HAVE_CUDA
+#include "cuda_operations.h"
+#endif
+
+namespace horovod {
+namespace common {
+
+// Table for storing Tensor metadata on rank zero. This is used for error
+// checking, stall checking and size calculations, as well as determining
+// when a reduction is ready to be done (when all nodes are ready to do it).
+using MessageTable = std::unordered_map<
+    std::string,
+    std::tuple<std::vector<MPIRequest>, std::chrono::steady_clock::time_point>>;
+
+class MPIChannel : public Channel {
```

Comment:
This [MPI]Channel concept feels very confusing - it seems to be combining some coordinator state, an abstraction over MPI for collective calls, and even abstraction for MPI ops.  Let's remove it for now and inline MPI everywhere to make this change simpler.

Reply:
I suppose the biggest downside of doing so is that we'd need to then couple the NCCL operations with MPI (as NCCL uses MPI ops at several points, e.g., hierarchical allreduce).  

In the current design the MPIChannel allows us to use composition in NCCL ops in place of inheritance, so we can sub out a different CPU channel as desired.  Do you feel it's preferable to have NCCL ops include MPI headers and have them call MPI operations directly?

Reply:
Removed channel concept.